blueprints/server-capabilitites.md

```diff
@@ -0,0 +1,162 @@
+  - Ability to add, delete Server Capability Types
+  - May be a text input, which adds to a list, with list element buttons to remove from the list
+
+
+### Traffic Ops Impact
+
+`/server_server_capabilities`
+  - List all Server capabilities
+  - GET+POST+DELETE
+  - PUT doesn't make sense. To remove one and add another, DELETE and POST.
+
+`/deliveryservices_required_capabilities`
+  - List all DS Required Capabilities
+  - GET+POST+DELETE
+  - PUT doesn't make sense. To remove one and add another, DELETE and POST.
+
+`/server_capabilities`
+  - List, create, and delete Server Capabilities.
+  - POST+GET+DELETE
+  - PUT doesn’t make sense, Capabilities are strings, if one is misspelled it should be removed and re-added. Simpler, less error-prone than cascading changes.
+
+Additionally, ATS config generation (which is currently in Traffic Ops, but in the process of being moved to its own library/app) will require changes. Primarily `parent.config`, as in the description, to prevent Mids being assigned as Parents for DSes for which they lack the capabilitites.
+
+#### REST API Impact
+
+See Traffic Ops Impact
+
+#### Client Impact
+
+See Traffic Ops Impact - client functions for each endpoint.
+
+#### Data Model / Database Impact
+
+New tables for Capabilitites: `server_capability`, `server_server_capability`, `delivery_service_required_server_capability`.
+
+### ORT Impact
+
+n/a
+
+### Traffic Monitor Impact
+
+n/a
+
+### Traffic Router Impact
+
+n/a
+
+### Traffic Stats Impact
+
+n/a
+
+### Traffic Vault Impact
+
+n/a
+
+### Documentation Impact
+
+The new UI and API endpoints will be documented. See Traffic Ops Impact.
+
+### Testing Impact
+
+The new UI, API endpoints, and ATS config changes, will have TO API tests and unit tests. See Traffic Ops Impact.
+
+
+### Performance Impact
+
+No performance impact expected.
+
+### Security Impact
+
+n/a
+
+### Upgrade Impact
+
+None. Upgrade will have no Required Capabilities, and config and server assignments will remain unchanged until an operator creates Server Capabilities.
+
+### Operations Impact
+
+None, until Server Capabilities are added. If a TC operator decides to use Server Capabilitites, they will they have to learn how to use them, and create and assign the Server Capabilitites they desire.
+
+### Developer Impact
+
+Code is added to TO endpoints, and config generation. Config generation complexity increases very slightly. TO endpoints are stand-alone and don't affect other endpoints.
+
+## Alternatives
+
+There are several other ways to do this.
+
+Within 'Server Capabilities' the API endpoints could be added to existing Server and Delivery Service endpoints. We are proposing they be standalone for now, because it's unclear how long-lived this feature will be, and how it will interact with 'Flexible Cachegroups' and other planned projects. Standalone endpoints give us more flexibility, especially around API versioning, to change or remove the feature in the future.
```

Comment:
Doesn't it take exactly the same deprecation and removal timeframe to get rid of a field on an object as it does to get rid of an endpoint?

Reply:
Sort of. But the existing one requires rolling version APIs, and doing versioning in the code. Where the standalone one, we can just leave it there and not do anything about it, until we remove it when we roll the API Major Version. Or, if we remove the actual feature from the server, we can just change the endpoints to return a 5xx. Where the existing endpoints can't just return a bad code, they'll have to ignore posted data, and then the user loses data they posted and got a success on. Likewise, the field has to be returned, so it's returned as a `null` or -1 or empty string, or some other data that may make the user think it actually exists.

You're right, it still has to follow the same rules. But separate endpoints are easier to deal with, easier in the API Versioning code, and removal has more options we can do and is less dangerous to the user.

Reply:
I'm still not sold, but I'm not going to hold anything up over it.